### PR TITLE
Add pg_post_event and DictProxy C API to release GIL, fix timer issues and speedups

### PR DESCRIPTION
**Pull request by ankith26**
A port of my ol PR https://github.com/pygame/pygame/pull/3062 (I am skipping re-copying the details here, the link should have all the necessary info)

It's been a while but I just remembered I had this PR lying around for a while and it happens to fix a nasty bug that has been reported a couple of times over the years (I noticed upstream also seems to have patched the bug, but without any of my changes)

Anyways this PR does other improvements apart from the fix (hence the large diff). MyreMylar approved this PR in the past, and RN nothing much has changed since (though I took in nthykier's reviews while re-making this PR as I said earlier on)

Personally I would consider this PR semi-important for the next release

### CODE REVIEW

File: src_c/_pygame.h [hunk 1]
```diff
@@ -42,6 +42,19 @@
 
 #include <SDL.h>
 
+/* DictProxy is useful for event posting with an arbitrary dict. Maintains
+ * state of number of events on queue and whether the owner of this struct
+ * wants this dict freed. This DictProxy is only to be freed when there are no
+ * more instances of this DictProxy on the event queue. Access to this is
+ * safeguarded with a per-proxy spinlock, which is more optiminal than having
```
Comment (Starbuck5) [inline src_c/_pygame.h:49]: typo: optiminal -> optimal

**Comment by Starbuck5** [comment]
Devlog of me wading through these changes:
I like the simplifications this seems to enable in the mixer and music code, it's a nice API to just say `pg_post_event`

Do you think this will have an impact that is possible to measure in user facing APIs? Since you're talking about this being an optimization. It seems like the main motivation is really to make things more convenient for the external callers. But the external callers could just call a function to get the GIL, doesn't seem so inconvenient.

I think I followed everything I've read so far, but I now need to go into the timer code, and that is a complex diff.

File: src_c/time.c [hunk 2]
```diff
@@ -28,128 +28,223 @@
...
+    PG_TIMER_NO_ERROR,
+    PG_TIMER_SDL_ERROR,
+    PG_TIMER_MEMORY_ERROR,
+} pgSetTimerErr;
+
+/* This is a doubly linked list of event timers. Access to this linked list
+ * must be protected with an SDL mutex.
+ * An SDL mutex would be redundant if python GIL was unconditionally held in
+ * the C callback function. But acquiring GIL can be costlier than acquiring a
+ * dedictated mutex for event timers, because the GIL is usually held more
```
Comment (Starbuck5) [inline src_c/time.c:42]: typo: dedicated

File: src_c/time.c [hunk 3]
```diff
@@ -300,50 +407,93 @@ time_set_timer(PyObject *self, PyObject *args, PyObject *kwargs)
...
+        case PG_TIMER_NO_ERROR:
+            Py_RETURN_NONE;
+        case PG_TIMER_SDL_ERROR:
+            return RAISE(pgExc_SDLError, SDL_GetError());
+        case PG_TIMER_MEMORY_ERROR:
+            return PyErr_NoMemory();
+        default:
+            return RAISE(
+                pgExc_SDLError,
+                "Unknown and unhandled internal error occured while handling "
```
Comment (Starbuck5) [inline src_c/time.c:493]: typo: occurred

**Comment by ankith26** [comment]
While fixing the typos you pointed out, I just realised that there is a bit of an inconsistency with this PR. This PR makes it clear that events posted by any of the event posting functions, post a shallow-copy of the event (the dict is a common reference) *however* this PR also adds special-casing for the empty dict which would mean any event posted with an empty dict is actually true-copied. I'm inclined to fixing this up and removing the special-case, I'm also pretty sure it's not going to significantly affect performance. However before I do that I'd like opinions on this, as this change would have user-facing impacts.

**Comment by ankith26** [comment]
While adding more event related tests, I also a small issue in pypy specific code, as it was a one-line fix I did it here.

**Pull request merged**